### PR TITLE
Add data plugin command entry point

### LINKED ISSUE #992: Add entry point for plugin data commands
Allow plugin authors to write cli commands, which can be called via `verdi data <plugin-data-type>`.

### PR DESCRIPTION
**Pull request by DropD**
close #992

### CODE REVIEW

**Comment by ltalirz** [comment]
@DropD very much looking forward to this one!
Let me know when it passes the test.

**Comment by DropD** [comment]
@ltalirz Tests are passing now, it was only a missing requirement.

**Comment by DropD** [comment]
I will add documentation with links to `click` docs and an example.

**Comment by DropD** [comment]
@ltalirz  please check this out, build the documentation and try to use Developers guide > Developer commandline plugin tutorial.

File: docs/source/developer_guide/devel_tutorial/cmdline_plugin.rst [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+   verdi data yourplugin-float --help
+
+It should print some basic usage information containing the docstring we gave to the ``float_cmd()`` function.
+
+The last step is now implementing ``verdi data yourplugin-float export`` command that exports our FloatData instance to a file.
+
+Append to file ``float_cmd.py``::
+
+   @float_cmd.command()
+   @click.option('--outfile', '-o', type=click.Path(dir_ok=False), help='write output to this file (by default print to stout).'
```
Comment (ltalirz) [inline docs/source/developer_guide/devel_tutorial/cmdline_plugin.rst]: should be dir_okay

File: docs/source/developer_guide/devel_tutorial/cmdline_plugin.rst [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+   verdi data yourplugin-float --help
+
+It should print some basic usage information containing the docstring we gave to the ``float_cmd()`` function.
+
+The last step is now implementing ``verdi data yourplugin-float export`` command that exports our FloatData instance to a file.
+
+Append to file ``float_cmd.py``::
+
+   @float_cmd.command()
```
[the remark this answers is not in the document]
Comment (DropD) [reply docs/source/developer_guide/devel_tutorial/cmdline_plugin.rst]: Nope, the command decorator uses the function name by default. I will add a paragraph explaining this.

File: docs/source/developer_guide/devel_tutorial/cmdline_plugin.rst (hunk 1 above)
Comment (DropD) [reply docs/source/developer_guide/devel_tutorial/cmdline_plugin.rst]: Agreed

File: docs/source/developer_guide/devel_tutorial/cmdline_plugin.rst [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+This test can now be run using ``TEST_BACKEND=django python -m unittest discover`` from your top level project directory ``aiida-yourplugin``.
+
+As a further exercise, try adding a ``--format`` option to choose between plain text and, say json.
+
+Understanding commandline plugins
+---------------------------------
+
+The discovery of plugins via entry points follows exactly the same mechanisms as all other plugin types.
+
+The possibility of pluging cli commands into each other is a feature of ``click`` a python library that greatly simplifies the task. You can find in-debth documentation here: `Click 6.0 docs <click_docs>`_.
```
[the remark this answers is not in the document]
Comment (DropD) [reply docs/source/developer_guide/devel_tutorial/cmdline_plugin.rst]: Fixed in 7cfd00a.

**Comment by DropD** [comment]
The documentation was also updated for the tutorial parts to line up in the side bar. To mark them better visually, all their titles now start with 'Tutorial:'. The old data plugin command part was marked as outdated with a `.. warning::`. And should be removed eventually. 
It might however be better to reimplement the examples in the new system first?

**Comment by ltalirz** [comment]
Marking the tutorials in the developer documentation more clearly is a good first step.

In my opinion, further reorganisation is badly needed - the aiida developer documentation contains an enormous amount of information that should be categorised (and quite a bit of which is outdated).
@giovannipizzi I'm happy to prepare some thoughts on this. Then, perhaps the two of us can at some point take 1-2 hours to go through the whole developer documentation together to agree on how to move things (+what to delete). After this, I can implement them.

**Comment by giovannipizzi** [comment]
@ltalirz ok!

**Pull request merged**